### PR TITLE
*: move config file option tidb_enable_auto_analyze to sysvar

### PR DESCRIPTION
**Pull request by Alkaagr81**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/33769

Problem Summary:

The option `tidb_enable_auto_analyze `  has historically been a config option. But based on requirements from cloud & PM it should instead be a sysvar scope Global.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [X] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [X] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
The option `tidb_enable_auto_analyze `  has historically been a config option. But based on requirements from cloud & PM it should instead be a sysvar scope Global.
```

### CODE REVIEW

**Comment by Alkaagr81** [comment]
cc @morgo

File: session/bootstrap.go [hunk 1]
```diff
@@ -617,11 +617,13 @@ const (
 	version89 = 89
 	// version90 converts enable-batch-dml to a sysvar
 	version90 = 90
+	// version92 converts run-auto-analyze to a sysvar
+	version92 = 92
```
Comment (morgo) [inline session/bootstrap.go:621]: I assume you picked v92 because [this PR](https://github.com/pingcap/tidb/pull/34644) is 91? 

We can't guarantee the merge order of PRs, so it is best to just pick the next number in sequence and then fix conflicts when they arise.

File: config/config.go [hunk 2]
```diff
@@ -597,7 +600,6 @@ type Performance struct {
 	TCPKeepAlive          bool    `toml:"tcp-keep-alive" json:"tcp-keep-alive"`
 	TCPNoDelay            bool    `toml:"tcp-no-delay" json:"tcp-no-delay"`
 	CrossJoin             bool    `toml:"cross-join" json:"cross-join"`
-	RunAutoAnalyze        bool    `toml:"run-auto-analyze" json:"run-auto-analyze"`
```
Comment (morgo) [inline config/config.go:600]: For the upgrade to work, you'll need to move this back to here (not in top level RunAutoAnalyze).

File: session/bootstrap.go [hunk 3]
```diff
@@ -1847,6 +1850,13 @@ func upgradeToVer90(s Session, ver int64) {
 	valStr := variable.BoolToOnOff(config.GetGlobalConfig().EnableBatchDML)
 	importConfigOption(s, "enable-batch-dml", variable.TiDBEnableBatchDML, valStr)
 }
+func upgradeToVer92(s Session, ver int64) {
+	if ver >= version92 {
+		return
+	}
+	valStr := variable.BoolToOnOff(config.GetGlobalConfig().RunAutoAnalyze)
```
Comment (morgo) [inline session/bootstrap.go:1857]:
```suggestion
	valStr := variable.BoolToOnOff(config.GetGlobalConfig().Performance.RunAutoAnalyze)
```

File: sessionctx/variable/sysvar.go [hunk 4]
```diff
@@ -652,6 +652,15 @@ var defaultSysVars = []*SysVar{
 			return nil
 		},
 	},
+	{Scope: ScopeGlobal, Name: TiDBEnableAutoAnalyze, Value: BoolToOnOff(DefTiDBEnableAutoAnalyze), skipInit: true, Type: TypeBool,
```
Comment (morgo) [inline sessionctx/variable/sysvar.go:655]: Skip init is for session scoped vars, and deprecated:
```suggestion
	{Scope: ScopeGlobal, Name: TiDBEnableAutoAnalyze, Value: BoolToOnOff(DefTiDBEnableAutoAnalyze), Type: TypeBool,
```

**Review by morgo — COMMENTED** [review]
LGTM, just one minor nit about skip-init.

**Review by morgo — APPROVED** [review]

File: sessionctx/variable/sysvar_test.go [hunk 5]
```diff
@@ -659,7 +659,7 @@ func TestSettersandGetters(t *testing.T) {
 			case TiDBEnableChangeMultiSchema, TiDBDDLReorgBatchSize,
 				TiDBMaxDeltaSchemaCount, InitConnect, MaxPreparedStmtCount,
 				TiDBDDLReorgWorkerCount, TiDBDDLErrorCountLimit, TiDBRowFormatVersion,
-				TiDBEnableTelemetry, TiDBEnablePointGetCache:
+				TiDBEnableTelemetry, TiDBEnablePointGetCache, TiDBEnableAutoAnalyze:
```
Comment (morgo) [inline sessionctx/variable/sysvar_test.go:662]: I tested this locally. You should be able to revert this change - `TiDBEnableAutoAnalyze` doesn't need to be mentioned here.

File: domain/domain.go [hunk 6]
```diff
@@ -1328,7 +1325,7 @@ func (do *Domain) UpdateTableStatsLoop(ctx sessionctx.Context) error {
 	}
 	do.SetStatsUpdating(true)
 	do.wg.Run(func() { do.updateStatsWorker(ctx, owner) })
-	if RunAutoAnalyze {
+	if variable.RunAutoAnalyze.Load() {
```
Comment (bb7133) [inline domain/domain.go:1328]: Changing the code here does not take effect actually, because `UpdateTableStatsLoop` is only called on the bootstrap stage of TiDB if you check the [code](https://github.com/pingcap/tidb/blob/ebb91121643c2a5e17c3c8af8f36b83d61688789/session/session.go/#L2939) in `BootstrapSession()`.

This is fine before because when it is changed in the configuration file, TiDB always needs to restart and the `BootstrapSession()` is called, but making it a sysvar means it the value can be changed dynamically, so:

* When it was `FALSE` and set to `TRUE`, the auto-analyze worker will not be invoked.
* When it was `TRUE` and set to `FALSE`, the auto-analyze worker will still be there.

You can easily verify it by, for example, adding a log in [`HandleAutoAnalyze()`](https://github.com/pingcap/tidb/blob/ebb91121643c2a5e17c3c8af8f36b83d61688789/statistics/handle/update.go/#L1037-L1038).

So my suggestion is removing the `if` here:

```
// always runs the loop
do.wg.Run(func() { do.autoAnalyzeWorker(owner) })
```
And moving `if` to [here](https://github.com/pingcap/tidb/blob/ebb91121643c2a5e17c3c8af8f36b83d61688789/domain/domain.go/#L1508) like:
```
		select {
		case <-analyzeTicker.C:
			if owner.IsOwner() && variable.RunAutoAnalyze.Load() {
				statsHandle.HandleAutoAnalyze(do.InfoSchema())
			}
		case <-do.exit:
			return
		}
```
So that the sysvar takes effect really.

@chrysan Please let me know if you have any comment on it, thanks!

File: statistics/integration_test.go [hunk 7]
```diff
@@ -499,7 +498,6 @@ func TestAnalyzeLongString(t *testing.T) {
 }
 
 func TestOutdatedStatsCheck(t *testing.T) {
-	domain.RunAutoAnalyze = false
```
Comment (bb7133) [inline statistics/integration_test.go:502]: Since the default value of `RunAutoAnalyze` is `true`, I think simply removing it may cause some trouble. @chrysan PTAL.

Comment (morgo) [reply statistics/integration_test.go:502]: It's false in the test suite. See the change in session/bootstrap.go:2012.

Comment (chrysan) [reply statistics/integration_test.go:502]: Prefer to keep the variable set in ut or at least an assertion to make the ut safe no matter what has been changed in variable module.

File: domain/domain.go (hunk 6 above)
Comment (chrysan) [reply domain/domain.go]: LGTM

File: statistics/integration_test.go (hunk 7 above)
Comment (bb7133) [reply statistics/integration_test.go:502]: Agreed, an explicit update for UT is more solid. @Alkaagr81 could you update this?

Comment (morgo) [reply statistics/integration_test.go:502]: The problem with setting it in the test is:
- If you update the atomic directly, it can get overwritten when the sysvar cache is updated. This is because the sysvar cache calls the SetGlobal func with the value from the mysql.global_variables.
- To safely change it, tk.MustExec("set global sysvar=x") must be used. But tk is not available in all of the tests. We could add it, but it's quite a bit more code to set up each time.

This is why it uses the bootstrap to disable auto-analyze instead.

Comment (morgo) [reply statistics/integration_test.go:502]: We can however add an assertion that runautoanalyze is false. Reading the atomic value is no problem.

Comment (bb7133) [reply statistics/integration_test.go:502]: I see, an alternative way is to create it as another 'enhancement' issue as a backlog, then we could make this PR delivered with the next version(v6.1)

**Comment by bb7133** [comment]
@Alkaagr81 Please resolve the conflict.

**Review by bb7133 — APPROVED** [review]
LGTM

File: statistics/integration_test.go (hunk 7 above)
Comment (morgo) [reply statistics/integration_test.go:502]: Forked to https://github.com/pingcap/tidb/issues/34792

**Comment by morgo** [comment]
As discussed, we will handle the upgrade step for this in https://github.com/pingcap/tidb/pull/34711
